control/state.py

```diff
@@ -7,109 +7,179 @@
+        key = self.NAMESPACE_PREFIX + subsystem_nqn + "_" + nsid
+        self._add_key(key, val)
+
+    def remove_namespace(self, subsystem_nqn: str, nsid: str):
+        """Removes a namespace from the state data store."""
+        key = self.NAMESPACE_PREFIX + subsystem_nqn + "_" + nsid
+        self._remove_key(key)
+
     def add_subsystem(self, subsystem_nqn: str, val: str):
-        pass
+        """Adds a subsystem to the state data store."""
+        key = self.SUBSYSTEM_PREFIX + subsystem_nqn
+        self._add_key(key, val)
 
-    @abstractmethod
     def remove_subsystem(self, subsystem_nqn: str):
-        pass
+        """Removes a subsystem from the state data store."""
+        key = self.SUBSYSTEM_PREFIX + subsystem_nqn
+        self._remove_key(key)
+
+        # Delete all keys related to subsystem
+        state = self.get_state()
+        for key in state.keys():
+            if (key.startswith(self.NAMESPACE_PREFIX + subsystem_nqn) or
+                    key.startswith(self.HOST_PREFIX + subsystem_nqn) or
+                    key.startswith(self.LISTENER_PREFIX + subsystem_nqn)):
+                self._remove_key(key)
 
-    @abstractmethod
     def add_host(self, subsystem_nqn: str, host_nqn: str, val: str):
-        pass
+        """Adds a host to the state data store."""
+        key = "{}{}_{}".format(self.HOST_PREFIX, subsystem_nqn, host_nqn)
+        self._add_key(key, val)
 
-    @abstractmethod
     def remove_host(self, subsystem_nqn: str, host_nqn: str):
-        pass
+        """Removes a host from the state data store."""
+        key = "{}{}_{}".format(self.HOST_PREFIX, subsystem_nqn, host_nqn)
+        self._remove_key(key)
 
-    @abstractmethod
-    def add_listener(self, subsystem_nqn: str, traddr: str, trsvcid: str,
-                     val: str):
-        pass
+    def add_listener(self, subsystem_nqn: str, gateway: str, trtype: str,
+                     traddr: str, trsvcid: str, val: str):
+        """Adds a listener to the state data store."""
+        key = "{}{}_{}_{}_{}_{}".format(self.LISTENER_PREFIX, gateway,
+                                        subsystem_nqn, trtype, traddr, trsvcid)
+        self._add_key(key, val)
 
-    @abstractmethod
-    def remove_listener(self, subsystem_nqn: str, traddr: str, trsvcid: str):
-        pass
+    def remove_listener(self, subsystem_nqn: str, gateway: str, trtype: str,
+                        traddr: str, trsvcid: str):
+        """Removes a listener from the state data store."""
+        key = "{}{}_{}_{}_{}_{}".format(self.LISTENER_PREFIX, gateway,
+                                        subsystem_nqn, trtype, traddr, trsvcid)
+        self._remove_key(key)
 
     @abstractmethod
     def delete_state(self):
+        """Deletes state data store."""
         pass
 
-    @abstractmethod
-    def restore(self, callbacks):
-        pass
+
+class LocalGatewayState(GatewayState):
+    """Persists gateway NVMeoF target state to a local object.
```

Comment:
I wouldn't use the word "persists" when it comes to in-memory state.  Also, "local object" is pretty confusing:
```suggestion
    """Records gateway NVMeoF target state in a dictionary.
```

Reply:
👍